src/hal/src/pso/mod.rs

```diff
@@ -229,3 +229,13 @@ pub enum Constant {
     F32(f32),
     F64(f64),
 }
+
+/// Pipeline state which may be static or dynamic.
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum State<T> {
```

Comment:
Yep, if this `State` approach looks good, I'd prefer to transition all of the baked states over too, otherwise I'll update this PR to move these values into `BakedStates`.

Reply:
Hmm, so this is the deciding point... I'll think about it till the end of the day.
There is just a few things to address here before we can merge, it seems.

Reply:
We could change them over in later PRs too.

IMO it does seem more elegant to group related states together (i.e. instead of spreading stencil flags across two separate fields), and more consistent with Vulkan I guess

Reply:
There are following choices:
  A. let all static states (that could be dynamic) to go into `BakedStates`
  B. have them as `Option` within bigger state blocks (e.g. stencil ref inside stencil face)
  C. have them as `State` within bigger state blocks

I think that 1) having states be more local helps (so, B and C are slightly better), and 2) this is a somewhat special mechanics of baking states in the pipeline, so it makes sense to have a separate enum type (`State`). Thus, I tend to agree with the direction this PR is taking.

However, @msiglreith please check this out and let us know in general if you have concerns.

Reply:
No concerns from my side. I personally prefer option C.